Add GoogleDocs shortcode

goldenapples opened the pull request:
#113

goldenapples commented:
_Still needs:_
- [ ] height and width attributes, should be available to all these document types
- [ ] JS hooks, to conditionally show the fields that are relevant for each document type.
- [ ] Investigating any other patterns of these document types or additional options that need to be supported
- [ ] Tests of each of these using the shortcake-bakery-responsive class to see if that makes sense on each document type.

goldenapples commented:
@danielbachhuber Want to give this a review and see if it covers all the expected use cases? I included all the embeddable Google Docs document types, except for "drawing", because that (a) doesn't fit the pattern the others do, and (b) just outputs an `<img>` tag which WP can handle well enough on its own.

danielbachhuber commented on inc/shortcodes/class-googledocs.php:
```diff
@@ -0,0 +1,245 @@
+						'form'          => esc_html__( 'Form', 'shortcake-bakery' ),
+						'map'           => esc_html__( 'Map', 'shortcake-bakery' ),
+					),
+					'description'  => esc_html__( 'Type of document to embed', 'shortcake-bakery' ),
```
Isn't the type of document based on the URL? What happens if I select `map` but enter a spreadsheet URL?

goldenapples replied:
Good point. The only real reason for this attribute is to wire up the JS hooks, so that we don't have to display all the fields specific to presentations on all the other document types and so on. But that could really be done just as easily by reading the document type from the URL field.

danielbachhuber replied:
> But that could really be done just as easily by reading the document type from the URL field.

:+1:

danielbachhuber commented on assets/js/src/shortcake-bakery-admin.js:
```diff
@@ -110,4 +110,32 @@ jQuery( document ).ready( function ( $ ) {
+	var gdocFields = {
+		all:          [ 'headers', 'start', 'loop', 'delayms', 'allowfullscreen' ],
+		spreadsheet:  [ 'headers' ],
+		presentation: [ 'start', 'loop', 'delayms', 'allowfullscreen' ]
```
Rather than hardcoding these values, can we produce a little bit of an abstraction? Maybe `display_if=>type=>spreadsheet`?

goldenapples replied:
Yeah, we've talked about that. Would it be good to work that into Shortcake itself? Or should we build out the API here first to validate it before moving it into Shortcake?

danielbachhuber replied:
> Or should we build out the API here first to validate it before moving it into Shortcake?

I dig this idea.

danielbachhuber commented on inc/shortcodes/class-googledocs.php:
```diff
@@ -0,0 +1,245 @@
+					'default'      => 3000,
+				),
+				array(
+					'label'        => esc_html__( 'Allow fullscreen mode?', 'shortcake-bakery' ),
```
Does this really need to be an option? When wouldn't you want fullscreen mode?

goldenapples replied:
shrug.. I don't know? I just figured it was an option in the embed code, so I'd expose it here as well.

danielbachhuber commented:
> Want to give this a review and see if it covers all the expected use cases?

Left some comments — looking pretty good! If this doesn't cover all use cases, I think it's a great start.

goldenapples commented:
Updated the reversal and callback in a few ways: 
* removed the "type" attribute, as the type can always be parsed out of the URL easily enough.
* removed the "allow full screen?" attribute from presentation

goldenapples commented:
Think this is ready for a #reviewmerge.

danielbachhuber commented on inc/class-shortcake-bakery.php:
```diff
@@ -139,6 +140,13 @@ public function action_admin_enqueue_scripts() {
+			'shortcodes' => array_flip(
+				array_map(
+					function( $class ) {
+						return str_replace( 'Shortcake_Bakery\\Shortcodes\\', '', $class );
```
Shouldn't we run `get_shortcode_tag()` for each class, as other namespaces won't benefit from this `str_replace()` ?

goldenapples replied:
This whole point here to to pass the shortcode tag into javascript; it passes an object that looks like `"GoogleDocs": "googledocs"` or whatever the tag names are filtered to - ie `"GoogleDocs": "fusion-googledocs"`...

I guess the `str_replace` isn't necessary, other than making the js object a little less messy looking. It'd work just as well to pass the full namespaced class as the keys of the object. I'll do that.

danielbachhuber commented on inc/shortcodes/class-googledocs.php:
```diff
@@ -0,0 +1,265 @@
+			case 'document':
+				return sprintf( '<iframe class="%s" src="%s/pub?embedded=true"%s%s frameborder="0" marginheight="0" marginwidth="0"></iframe>',
+					esc_attr( $iframe_classes ),
+					esc_url_raw( $attrs['url'] ),
```
`esc_url()`

danielbachhuber commented on inc/shortcodes/class-googledocs.php:
```diff
@@ -0,0 +1,265 @@
+				);
+				return sprintf( '<iframe class="%s" src="%s" %s%sframeborder="0" marginheight="0" marginwidth="0" allowfullscreen="true" mozallowfullscreen="true" webkitallowfullscreen="true"></iframe>',
+					esc_attr( $iframe_classes ),
+					esc_url_raw( $url ),
```
`esc_url()`

danielbachhuber commented on inc/shortcodes/class-googledocs.php:
```diff
@@ -0,0 +1,265 @@
+				);
+				return sprintf( '<iframe class="%s" src="%s" %s%sframeborder="0" marginheight="0" marginwidth="0">%s</iframe>',
+					esc_attr( $iframe_classes ),
+					esc_url_raw( $url ),
```
`esc_url()`

danielbachhuber commented on inc/shortcodes/class-googledocs.php:
```diff
@@ -0,0 +1,265 @@
+			case 'map':
+				return sprintf( '<iframe class="%s" src="%s" %s%sframeborder="0" marginheight="0" marginwidth="0"></iframe>',
+					esc_attr( $iframe_classes ),
+					esc_url_raw( $attrs['url'] ),
```
`esc_url()`

danielbachhuber commented on inc/shortcodes/class-googledocs.php:
```diff
@@ -0,0 +1,265 @@
+					wp_kses_one_attr( $width_attr, 'img' ),
+					wp_kses_one_attr( $height_attr, 'img' )
+				);
+		}
```
Could we have some abstraction so we only need to produce the `iframe` once?

goldenapples replied:
Yup, updated in fbbce4d.

danielbachhuber commented on tests/test-googledocs-shortcode.php:
```diff
@@ -0,0 +1,99 @@
+		);
+	}
+
+	private function expect_reversal( $embed, $reversal ) {
```
I'm not a huge fan of this abstraction because I think it decreases the readability of the tests. But, not worth changing at this point.

danielbachhuber commented:
@goldenapples few smaller changes to make

goldenapples commented:
Updated in response to comments, let me know how this looks?

danielbachhuber commented on inc/shortcodes/class-googledocs.php:
```diff
@@ -0,0 +1,270 @@
+					return $attribute_string . sanitize_key( $attr_key ) . '="' . esc_attr( $additional_attributes[ $attr_key ] ) . '" ';
+				} ) : '' ) . '>' .
+			( $inner_content ? esc_html( $inner_content ) : '' ) .
+			'</iframe>';
```
Oy.
